scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,169 @@
 package org.scalafmt.sbt
-import scala.meta.internal.tokenizers.PlatformTokenizerCache
-import sbt._
-import Keys._
+
+import org.scalafmt.{Formatted, Scalafmt}
+import org.scalafmt.config.{Config, ScalafmtConfig, ScalafmtRunner}
+import sbt.Keys._
+import sbt.{Def, _}
+import complete.DefaultParsers._
+import sbt.util.Logger
+
+import scala.util.{Failure, Success, Try}
 
 object ScalafmtPlugin extends AutoPlugin {
   override def trigger: PluginTrigger = allRequirements
+
   object autoImport {
-    val scalafmt: Command =
-      Command.args("scalafmt", "run the scalafmt command line interface.") {
-        case (state, args) =>
-          org.scalafmt.cli.Cli.main("--non-interactive" +: args.toArray)
-          PlatformTokenizerCache.megaCache.clear()
-          state
+    lazy val scalafmt = TaskKey[Unit]("scalafmt", "Format Scala sources")
+    lazy val scalafmtCheck = TaskKey[Boolean](
+      "scalafmt-check",
+      "Check that Scala sources is formatted properly")
+    lazy val scalafmtOnCompile =
+      SettingKey[Boolean]("scalafmt-on-compile", "Format source when compiling")
+    lazy val scalafmtConfig =
+      TaskKey[File]("scalafmt-config", "Scalafmtter config file")
+    lazy val scalafmtSbt = TaskKey[Unit]("scalafmt-sbt", "Format SBT sources")
+    lazy val scalafmtSbtCheck = TaskKey[Boolean](
+      "scalafmtSbtCheck",
+      "Check that SBT sources is formatted properly")
+    lazy val scalafmtOnly =
+      InputKey[Unit]("scalafmt-only", "Format only given file")
+  }
+  import autoImport._
+
+  private lazy val scalaConfig =
+    scalafmtConfig.map(Config.fromHoconFile(_).get)
+  private lazy val sbtConfig =
+    scalaConfig.map(conf => conf.copy(runner = conf.runner.forSbt))
+
+  private type Input = String
+  private type Output = String
+
+  private def withFormattedSources[T](
+      sources: Seq[File],
+      config: ScalafmtConfig
+  )(
+      onError: (File, Throwable) => T,
+      onFormat: (File, Input, Output) => T
+  ): Seq[Option[T]] = {
+    sources.map(
```

Comment:
This formats files sequentially, for large projects it may pay off to do `.par.map`, but I'll leave it to you to decide.

Reply:
we'll get parallelism for free across projects / configs because of the way sbt works, it's not so clear that parallelism at this level will improve performance. let's measure first.

Reply:
right, since  the plugin works at the granularity of project/config then you already get parallelism there. In the cli we use `.par` to run on all files in the repo